### PR TITLE
Allow updates/patches to pod disruption budgets

### LINKED ISSUE #45398: PodDisruptionBudget updates are forbidden
_PodDisruptionBudget_ is immutable at this moment. It would be a nice improvement to allow changes to it, at the least to the _minAvailable_ field. Not sure whether this counts as a _bug_ or a _feature request_ :)

After creating a _PodDisruptionBudget_ with the following manifest.

```yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  name: elasticsearch-master
  labels:
    app: elasticsearch
    role: master
spec:
  selector:
    matchLabels:
      app: elasticsearch
      role: master
  minAvailable: 2
```

When changing the spec to let's say `minAvailable: 3` and re-running `kubectl apply` it returns the following error:

```
The PodDisruptionBudget "elasticsearch-master" is invalid: spec: Forbidden: updates to poddisruptionbudget spec are forbidden.
```

**Kubernetes version**:

```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.2", GitCommit:"477efc3cbe6a7effca06bd1452fa356e2201e1ee", GitTreeState:"clean", BuildDate:"2017-04-19T20:33:11Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.6", GitCommit:"114f8911f9597be669a747ab72787e0bd74c9359", GitTreeState:"clean", BuildDate:"2017-03-28T13:36:31Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Google Container Engine 1.6.2 with cos image**:
- **OS** (e.g. from /etc/os-release):

```
BUILD_ID=9000.84.2
NAME="Container-Optimized OS"
GOOGLE_CRASH_ID=Lakitu
VERSION_ID=56
BUG_REPORT_URL=https://crbug.com/new
PRETTY_NAME="Container-Optimized OS from Google"
VERSION=56
GOOGLE_METRICS_PRODUCT_ID=26
HOME_URL="https://cloud.google.com/compute/docs/containers/vm-image/"
ID=cos
```

- **Kernel** (e.g. `uname -a`):

```
Linux gke-production-europe-we-auto-scaling-917da0af-zzzj 4.4.21+ #1 SMP Fri Feb 17 15:34:45 PST 2017 x86_64 Intel(
R) Xeon(R) CPU @ 2.50GHz GenuineIntel GNU/Linux
```

**What happened**:

Changing the value of `spec.minAvailable` in a _PodDisruptionBudget_ failed due to updates being forbidden.

**What you expected to happen**:

At least `spec.minAvailable` being allowed to change because it's often used in conjunction with a _StatefulSet_ which is likely to grow over time by incrementing the number of replicas, at which time you also want to grow _minAvailable_ in the _PodDisruptionBudget_.

### PR DESCRIPTION
**Pull request by davidmccormick**
**What this PR does / why we need it**:  It removes the immutability of pod disruption budgets.  Encouraging teams to use Pod Disruption Budgets is a vital part of us allowing us to effectively perform maintenance on our clusters without inadvertently taking out one of the services running within it.  With the present Pod Disruption Budgets being immutable this presents problems incorporating them into application charts (neither kubectl with force or helm can properly handle changing a pdb object - forcing an administrator to delete and re-create it).  It has been said in the issue that there is no longer any design reason requiring pdbs to be immutable, so let's remove this control and allow them to be patched.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #45398
https://github.com/kubernetes/kubernetes/issues/45398
**Special notes for your reviewer**:
I have removed the DeepEqual check and then updated the tests in line with updates being allowed.  In updating the tests I found that the order two of the test parameters had been accidentally transposed so I have corrected that and updated them to show that updates are now successful but pdb validation still applies - i.e. you are still not allowed to mix MaxUnavailable and MinAvailable values.

**Release note**:

```release-note
Allow updates/patches to pod disruption budgets
```

### CODE REVIEW

**Comment by neolit123** [comment]
@davidmccormick 
please add a release note explaining the change.

**Comment by davidmccormick** [comment]
ok have added a release note - does it look ok?

**Comment by neolit123** [comment]
@davidmccormick 
normally RN are kept down to a sentence or two.
the title seems like a good one.
> Allow updates/patches to pod disruption budgets

the tool that collects release notes also inserts a PR link next to them for better context.

i would defer to the maintainers for further comments on the PR itself.
thanks.

**Comment by davidmccormick** [comment]
ok thanks, have updated the release note as suggested

**Comment by davidmccormick** [comment]
@hongchaodeng @janetkuo can you review this PR please?

**Comment by krmayankk** [comment]
@kubernetes/sig-apps-feature-requests  @kubernetes/sig-apps-pr-reviews  can some review or comment. This seems like we should do ?

**Comment by enisoc** [comment]
@davidmccormick Sorry for the delay. Are you able to pick this back up? I personally want to see this happen too and I'll volunteer to be a reviewer and help track down an approver. If you don't have time to continue this, please let us know so we can open a new PR.

**Review by enisoc — COMMENTED** [review]
I think we should have some integration/e2e tests for scenarios involving mutation of PDB specs. Maybe one [e2e](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/test/e2e/apps/disruption.go) test for a common use case like "test PDB works for maxUnavailable=N, then change to N+1 and check again". In [integration](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/test/integration/evictions/evictions_test.go), we can test more specific cases like "put a matching Pod in DisruptedPods list, then change label selector so that Pod no longer matches, and make sure the controller removes it from the list".

Do you have time to work on those?

File: pkg/apis/policy/validation/validation_test.go
Comment (davidmccormick) [reply pkg/apis/policy/validation/validation_test.go]: Yes, I think you are right - this is wrong and doesn't really affect the validation or the test cases.

**Comment by davidmccormick** [comment]
Hi @enisoc, thanks so much for picking this up!!  I am happy to work on this, I don't have any experience of the integration or e2e test suites but I'm happy to learn if you don't mind providing some guidance?  A colleague of mine has some experience here and so he's giving me a bit of intro to them.

**Comment by enisoc** [comment]
For e2e, I recommend starting by making a copy of this bare-bones test case and place it just below:

https://sourcegraph.com/github.com/kubernetes/kubernetes@7693a1d5fe2a35b6e2e205f03ae9b3eddcdabc6b/-/blob/test/e2e/apps/disruption.go#L56-58

Then look at the body of the table-based test below, and start pulling out some additional steps you'll need:

https://sourcegraph.com/github.com/kubernetes/kubernetes@7693a1d5fe2a35b6e2e205f03ae9b3eddcdabc6b/-/blob/test/e2e/apps/disruption.go#L148-216

The test I'm envisioning would start out similarly:

1. Create Pods and RS.
2. Create PDB.
3. Find a running Pod.
4. Try to evict that Pod.
5. Check that the eviction is denied.

But then you would add a second phase:

6. Update the PDB in-place to change minAvailable so the eviction will be allowed.
7. Try the eviction again on the same Pod. You might need to retry with a timeout, since the change will take effect asynchronously.
8. Check that the eviction eventually succeeds within a reasonable time (30s?).

The easiest way to run e2e is to push to your PR branch, but the round-trip time is very long. You can find instructions for running e2e locally here:

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/e2e-tests.md

Once we're on track for e2e we can talk about integration tests. Let me know if you need more pointers. Good luck and thanks for helping to fix this for all users!

**Comment by davidmccormick** [comment]
@enisoc Hi, apologies on the delay in getting this work done!  I have now got a bit more up-to-speed with the e2e test suit and have written an e2e test that covers the scenario you suggested.  Does it look ok to you?

You should be able to run the test on its own: -
```
kubetest --provider=local --test --test_args="--minStartupPods=1 --ginkgo.focus=UpdatingPodDisruptionBudgets"
```

File: test/e2e/apps/disruption.go
Comment (enisoc) [inline test/e2e/apps/disruption.go]: Since this returns the error, let's leave it up to the caller to check it. Otherwise we're handling it twice.

File: test/e2e/apps/disruption.go
Comment (enisoc) [inline test/e2e/apps/disruption.go]: You can remove this line now.

File: test/e2e/apps/disruption.go
Comment (enisoc) [inline test/e2e/apps/disruption.go]: I don't think we need to create a new `Describe`. Let's just add new `It`s inside the existing `SIGDescribe("DisruptionController", ...`.

File: test/e2e/apps/disruption.go
Comment (enisoc) [inline test/e2e/apps/disruption.go]: This line isn't needed.

File: test/e2e/apps/disruption.go
Comment (enisoc) [inline test/e2e/apps/disruption.go]: Conventionally, we prefer to pass around pointers to Pods and other API objects. Shallow copies cause unexpected semantics since there are often nested pointers.

File: test/e2e/apps/disruption.go
Comment (enisoc) [inline test/e2e/apps/disruption.go]: Like the case in the table-based test where `c.shouldDeny` is `false`, we can speed this up by retrying the eviction instead of using a big, fixed sleep:

https://sourcegraph.com/github.com/kubernetes/kubernetes@7693a1d5fe2a35b6e2e205f03ae9b3eddcdabc6b/-/blob/test/e2e/apps/disruption.go#L204-214

File: test/e2e/apps/disruption.go
Comment (enisoc) [inline test/e2e/apps/disruption.go]: This could make the test flaky, since the Update can fail due to a ResourceVersion error if it races with other updates such as the controller updating status (we use optimistic concurrency).

Here's an example of using the RetryOnConflict helper to avoid this:

https://sourcegraph.com/github.com/kubernetes/kubernetes@7693a1d/-/blob/test/e2e/framework/service_util.go#L1071-1089

**Review by enisoc — COMMENTED** [review]
Mostly there on the e2e! Just some mechanical issues.

**Comment by liggitt** [comment]
> There can be races between eviction requests and PDB mutations, but I would argue that allowing mutations does not increase the risk any more than we already accept just because PDBs can be deleted and recreated.

I agree. Very glad to see movement on this.

File: test/e2e/apps/disruption.go
Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: Thanks, was useful whilst developing the tests but sure can be removed now.

File: test/e2e/apps/disruption.go
Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: removed

File: test/e2e/apps/disruption.go
Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: Ok, how about I move it above the test for the successful eviction instead?  That way I am ensuring that the pods are in the correct state before I attempt the successful eviction.

File: test/e2e/apps/disruption.go
Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: changing to pass the pod by pointer

File: test/e2e/apps/disruption.go
Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: ok removed

File: test/e2e/apps/disruption.go
Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: ok, have taken out the waits and retry the eviction after checking we have the correct number of pods running

**Comment by davidmccormick** [comment]
Hmm by taking out the time.Sleep I now occasionally get a race error when running my test stating that my PDB is not ready when I try an eviction...

```
  Allows updates to pdb's [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apps/disruption.go:205

  Expected error:
      <*errors.StatusError | 0xc0018c55f0>: {
          ErrStatus: {
              TypeMeta: {Kind: "", APIVersion: ""},
              ListMeta: {SelfLink: "", ResourceVersion: "", Continue: ""},
              Status: "Failure",
              Message: "Cannot evict pod as it would violate the pod's disruption budget.",
              Reason: "TooManyRequests",
              Details: {
                  Name: "",
                  Group: "",
                  Kind: "",
                  UID: "",
                  Causes: [
                      {
                          Type: "DisruptionBudget",
                          Message: "The disruption budget foo is still being processed by the server.",
                          Field: "",
                      },
                  ],
                  RetryAfterSeconds: 0,
              },
              Code: 429,
          },
      }
      Cannot evict pod as it would violate the pod's disruption budget.
  not to have occurred
```

**Comment by davidmccormick** [comment]
> Hmm by taking out the time.Sleep I now occasionally get a race error when running my test stating that my PDB is not ready when I try an eviction...
> 
> ```
>   Allows updates to pdb's [It]
>   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apps/disruption.go:205
> 
>   Expected error:
>       <*errors.StatusError | 0xc0018c55f0>: {
>           ErrStatus: {
>               TypeMeta: {Kind: "", APIVersion: ""},
>               ListMeta: {SelfLink: "", ResourceVersion: "", Continue: ""},
>               Status: "Failure",
>               Message: "Cannot evict pod as it would violate the pod's disruption budget.",
>               Reason: "TooManyRequests",
>               Details: {
>                   Name: "",
>                   Group: "",
>                   Kind: "",
>                   UID: "",
>                   Causes: [
>                       {
>                           Type: "DisruptionBudget",
>                           Message: "The disruption budget foo is still being processed by the server.",
>                           Field: "",
>                       },
>                   ],
>                   RetryAfterSeconds: 0,
>               },
>               Code: 429,
>           },
>       }
>       Cannot evict pod as it would violate the pod's disruption budget.
>   not to have occurred
> ```

I've tried putting in a check for the pdb having been processed rather than the arbitrary wait, and all of the tests are now passing reliably - although I am not 100% sure that by doing this I haven't invalidated any of the tests.

File: pkg/apis/policy/validation/validation_test.go
Comment (krmayankk) [inline pkg/apis/policy/validation/validation_test.go]: why is MaxUnavailable removed ?

File: test/e2e/apps/disruption.go
Comment (krmayankk) [inline test/e2e/apps/disruption.go]: nit: should we instead use PDB's ?

File: test/e2e/apps/disruption.go
Comment (krmayankk) [inline test/e2e/apps/disruption.go]: move this definition of `e` near Evict ?

File: test/e2e/apps/disruption.go
Comment (krmayankk) [inline test/e2e/apps/disruption.go]: what is the difference between waitForPodsOrDie and locateARUnningPod

File: pkg/apis/policy/validation/validation_test.go
Comment (enisoc) [reply pkg/apis/policy/validation/validation_test.go]: It's invalid to set both `MinAvailable` and `MaxUnavailable`, so this test case should not have been trying to do so (it's not intended to test an error condition). The flawed test case incidentally used to work because we did not validate `.spec` at all during updates, since validating spec on creation and enforcing immutability made that unnecessary.

File: test/e2e/apps/disruption.go
Comment (enisoc) [reply test/e2e/apps/disruption.go]: If you want to get super nit-picky, the plural of an initialism shouldn't have an apostrophe, unless the initialism is written with a period after each letter. So in this case it should be "PDBs".

File: test/e2e/apps/disruption.go
Comment (enisoc) [reply test/e2e/apps/disruption.go]: "Wait for pods" means let everything come to equilibrium. "Locate a running Pod" means choose a victim to try to evict.

**Review by enisoc — APPROVED** [review]
/lgtm

File: test/e2e/apps/disruption.go
Comment (krmayankk) [reply test/e2e/apps/disruption.go]: thanks, just looked at the code. Since we are evicting after waiting for all pods to be running, we can simply just waitForPodsOrDie and then evict because we just tested all are running. waitForPods could return one of the running pods to be evicted. Not a must do optimization , but prevents listing the pods twice.

File: test/e2e/apps/disruption.go
Comment (oomichi) [reply test/e2e/apps/disruption.go]: +1

File: test/e2e/apps/disruption.go
Comment (oomichi) [reply test/e2e/apps/disruption.go]: How about
`should block and allow evicting pod with PodDisruptionBudget`
like other tests in this test module?

File: test/e2e/apps/disruption.go
Comment (oomichi) [inline test/e2e/apps/disruption.go]: Here could be
`"Waiting for the pdb to be processed in namespace %s", ns)`

File: test/e2e/apps/disruption.go
Comment (oomichi) [inline test/e2e/apps/disruption.go]: nit: We can use `framework.ExpectNoError(err)` here like line 400.

File: test/e2e/apps/disruption.go
Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: ok changing

File: test/e2e/apps/disruption.go
Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: ok will change

File: test/e2e/apps/disruption.go
Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: Having had a think about this I'd rather not combine these two operations (waiting for test setup equilibrium and selecting a pod for eviction) as I feel they are sufficiently different in intent and useful separately - although they do mean that the API is called twice.

File: test/e2e/apps/disruption.go
Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: moved as requested

**Comment by davidmccormick** [comment]
Hi @enisoc, would you be able to assist me?  The e2e tests are building and running fine on a linux VM but the bazel build is failing with: -

```
W0409 13:48:14.976] ERROR: /workspace/k8s.io/kubernetes/test/e2e/apps/BUILD:8:1: GoCompile test/e2e/apps/linux_amd64_stripped/go_default_library%/k8s.io/kubernetes/test/e2e/apps.a failed (Exit 1) compile failed: error executing command
```

Is there a way I can get more debug as to why it is failing to compile during the bazel build?

**Comment by davidmccormick** [comment]
@enisoc would you be so kind as to give this another look and re-apply your lgtm if you are happy with it?

**Review by oomichi — COMMENTED** [review]
The e2e test part seems good for me, thanks.

File: test/e2e/apps/disruption.go
Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: Hi, perhaps but it has been passed as a static "foo" in all of the existing tests so I've tried to keep it consistent with them.

File: test/e2e/apps/disruption.go
Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: Do you feel that this code is acceptable?

Comment (oomichi) [reply test/e2e/apps/disruption.go]: Yes, looks good for me.

**Comment by mortent** [comment]
/lgtm
Looks like the PR needs a rebase. You should also squash the commits before merge.

**Comment by davidmccormick** [comment]
@mortent hi, I've rebased against master and squashed my commits

File: pkg/apis/policy/validation/validation.go
Comment (liggitt) [inline pkg/apis/policy/validation/validation.go]: what is the purpose of setting and restoring the generation field? it seems like a no-op

Comment (liggitt) [reply pkg/apis/policy/validation/validation.go]: if possible, we should drop that behavior... validation shouldn't mutate objects, even temporarily

**Comment by liggitt** [comment]
one question about cleaning up the generation set/restore in ValidatePodDisruptionBudgetUpdate, lgtm otherwise

**Comment by davidmccormick** [comment]
Hi @liggitt, I couldn't work out why the pdb validation wanted to save and restore the generation number, so I've removed it and re-run the tests and still behaving as expected.  Looking better?

File: test/e2e/apps/disruption.go
Comment (liggitt) [inline test/e2e/apps/disruption.go]: won't this make the test flaky?

**Comment by liggitt** [comment]
validation change lgtm.

one comment on dropping the timeout on the test.

I didn't look at the new test, will defer to @kubernetes/sig-apps-pr-reviews for that

File: test/e2e/apps/disruption.go
Comment (krmayankk) [inline test/e2e/apps/disruption.go]: is the blocking if PDB doesnt allow it , a new behavior ?

Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: It’s the standard behavior of the pdb to block an eviction unless it’s conditions have been satisfied.

File: test/e2e/apps/disruption.go
Comment (davidmccormick) [reply test/e2e/apps/disruption.go]: I thought that the time was to allow the pod disruption budget to be processed.  The difference is that now we specifically wait for the pdb to be processed when we create or update the pdb in the waitForPdbToBeProcessed() function.

**Comment by davidmccormick** [comment]
had to rebase @mortent @liggitt @kow3ns can you re-approve please?

File: test/e2e/apps/disruption.go
Comment (liggitt) [reply test/e2e/apps/disruption.go]: > The difference is that now we specifically wait for the pdb to be processed when we create or update the pdb in the waitForPdbToBeProcessed() function.

That indicates the pdb controller has observed the update. It does not indicate the informer in the apiserver process (used by the eviction REST handler) has observed the update. In practice, the in-process informer is likely to always see the update events first, but that is not guaranteed.

Please keep an eye on this test post-merge to see if this increases flakiness.

**Comment by pc-rshetty** [comment]
will this be backported to 1.13 and 1.14 ?

**Comment by liggitt** [comment]
No, backports are limited to critical bug fixes. Features are not backported.

**Pull request merged**
